File: 452-minimum-number-of-arrows-to-burst-balloons/minimum-number-of-arrows-to-burst-balloons.cpp
class Solution {
    using pp=pair<int,int>;
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        sort(points.begin(),points.end());
        priority_queue<pp,vector<pp>,greater<pp>>pq;

        int n=points.size(),i=0;
        int cnt=0;
        int start=INT_MIN;

        while(i<n || !pq.empty()){
            if(pq.empty()){
                pq.push({points[i][0],points[i][1]});
                start=max(start,pq.top().second);
                i++;
            }
            while(i<n && points[i][0]<=start){
                pq.push({points[i][0],points[i][1]});
                i++;
            }
            while(!pq.empty() && pq.top().first<=start){
                start=min(start,pq.top().second);
                pq.pop();
            }
            if(!pq.empty()){
                start=max(start,pq.top().second);
            }
            cnt++;
        }
        return cnt;
    }
};

File: 452-minimum-number-of-arrows-to-burst-balloons/minimum-number-of-arrows-to-burst-balloons.java
class Solution {
    public int findMinArrowShots(int[][] points) {
        Arrays.sort(points,(a,b)->{
            if(Integer.compare(a[1],b[1])==0){
                return Integer.compare(a[0],b[0]);
            }
            return Integer.compare(a[1],b[1]);
        });
        // for(int i=0;i<points.length;i++){
        //     System.out.println(points[i][0] + ","+points[i][1]);
        // }
        int ans=1;
        int n=points.length;
        int currMaxW=points[0][1];
        int ind=1;

        while(ind<n){
            int currInd=ind;
            while(currInd<n && points[currInd][0]<=currMaxW){
                currInd++;
            }
            if(currInd<n){
                ans++;
                currMaxW=points[currInd][1];
            }
            ind=currInd+1;
        }

        return ans;
    }
}